ENH: Enable indexing with nullable Boolean

dsaxton opened the pull request:
closes #31503 

- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I think from the discussion in https://github.com/pandas-dev/pandas/issues/31503 that this is something people want to allow.

Commit:
Add test

Commit:
Remove test that checks for error

Commit:
Add frame test

Commit:
Don't raise with nullable boolean

Commit:
Don't modify result

Commit:
Add frame test

Commit:
Update whatsnew

Commit:
Fill NA

Commit:
Merge branch 'master' into bool-idx

Commit:
Remove some more tests

Commit:
Delete another test

TomAugspurger commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -212,6 +212,7 @@ Sparse
 ExtensionArray
 ^^^^^^^^^^^^^^
 
+- Enabled filtering with nullable Boolean arrays. (:issue:`31503`)
```
This will probably merit it's own section under "enhancements".

TomAugspurger commented on pandas/core/indexing.py:
```diff
@@ -2222,6 +2224,12 @@ def check_bool_indexer(index: Index, key) -> np.ndarray:
             )
         result = result.astype(bool)._values
+    elif is_extension_array_dtype(key) and is_bool_dtype(key):
+        mask = isna(key)
```
This should likely be `key.to_numpy(fill_value=False)`.

I think that will avoid the full scans of the values from the `isna` and `.any()`.

dsaxton replied:
Yes, that's much nicer

TomAugspurger commented on pandas/tests/extension/base/getitem.py:
```diff
@@ -158,22 +158,6 @@ def test_getitem_boolean_array_mask(self, data):
         result = pd.Series(data)[mask]
         self.assert_series_equal(result, expected)
 
-    def test_getitem_boolean_array_mask_raises(self, data):
```
Rather than removing, can you adapt this test to treat the NAs as False?

TomAugspurger commented on pandas/tests/indexing/test_na_indexing.py:
```diff
@@ -59,21 +59,3 @@ def test_series_mask_boolean(values, dtype, mask, box_mask, frame):
-
-
-@pytest.mark.parametrize("frame", [True, False])
-def test_indexing_with_na_raises(frame):
```
Same: adapt for treating NA as false.

TomAugspurger reviewed (commented):
Thanks for working on this. I think the tests you've deleted should all be updated instead.

Commit:
Use to_numpy

Commit:
Update whatsnew

Commit:
Don't check for NA

Commit:
Revert "Remove test that checks for error"

This reverts commit 9c5b9f06f54e570134a846dfde488c5c7b720671.

Commit:
Update NA test

Commit:
Revert "Remove some more tests"

This reverts commit f6e9ce598ffa103159b9a6b6fb3b13a2730ebc0c.

Commit:
Update Categorical test

Not sure how applicable the CategoricalIndex part is
here so going to remove for now

Commit:
Update getitem tests

dsaxton commented on pandas/tests/series/indexing/test_boolean.py:
```diff
@@ -70,14 +70,6 @@ def test_getitem_boolean_object(string_series):
     s2[mask] = 5
     tm.assert_series_equal(cop, s2)
 
-    # nans raise exception
```
This one was checking that an _object_ indexing array containing `bool` and `NaN` values raises, which I think should still be the case even if a nullable boolean does not?

TomAugspurger replied:
That sounds right.

Commit:
Update indexers.py

dsaxton commented on pandas/core/indexers.py:
```diff
@@ -392,10 +390,17 @@ def check_array_indexer(array: AnyArrayLike, indexer: Any) -> Any:
+        if is_extension_array_dtype(dtype):
+            indexer = indexer.to_numpy(dtype=bool, na_value=False)
+        else:
+            try:
```
Is this try / except still needed after the conversion above? I think we can always do `np.asarray(dtype=bool)` if we make it here?

Commit:
tm -> self

Commit:
Assert for EA not DataFrame

Commit:
Don't try / except

Commit:
Change check_indexer test

Removed the test on [True, False, None] which is no longer raising.
Not clear to me what the correct behavior should be, but this is now
returning a pandas array with pd.NA instead of None.

dsaxton commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -929,3 +929,23 @@ def test_diff():
     tm.assert_series_equal(result, expected)
+
+
+def test_nullable_boolean_mask_series():
```
Should these tests be removed in favor of the modified tests? May be a bit redundant with the others checking for the same / similar behavior.

jreback replied:
if they are truly redudant

Commit:
Modify __getitem__ for datetimelike

Commit:
Add back ValueError for non-boolean with NA

Commit:
Revert "Delete another test"

This reverts commit 12344071edfcbe84007457ad0f812fe367dbc67b.

Commit:
Fixup error message

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,11 @@ For example:
+Indexing with Nullable Boolean Arrays
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Previously indexing with a nullable Boolean array would raise a ``ValueError``, however this is now permitted with ``NA`` being treated as ``False``. (:issue:`31503`)
```
add an example of the before and the after (style this after similar notes in 1.0.0)

jreback commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -520,7 +521,11 @@ def __getitem__(self, key):
             # first convert to boolean, because check_array_indexer doesn't
             # allow object dtype
-            key = np.asarray(key, dtype=bool)
+            if is_extension_array_dtype(key):
```
hmm, not sure you actually need to do this here; as you already patched check_array_indexer which does this.

jreback commented on pandas/core/indexing.py:
```diff
@@ -2222,6 +2224,8 @@ def check_bool_indexer(index: Index, key) -> np.ndarray:
                 "the indexed object do not match)."
             )
         result = result.astype(bool)._values
+    elif is_extension_array_dtype(key) and is_bool_dtype(key):
```
why do we have this at all anymore? is check_array_indexer not sufficient ?

dsaxton replied:
Yeah good point, there's a lot of duplicate code happening. Will work on cleaning this up.

Commit:
Add before and after examples

Commit:
Get rid of some tests

Commit:
Cast another way

Commit:
Import

Commit:
Don't import unused

Commit:
Merge branch 'master' into bool-idx

jreback commented:
so since indexing for these's EAs is experimental, I would be ok with doing this for 1.0.2.

@TomAugspurger @jorisvandenbossche

TomAugspurger commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,31 @@ For example:
+Indexing with Nullable Boolean Arrays
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Previously indexing with a nullable Boolean array would raise a ``ValueError``, however this is now permitted with ``NA`` being treated as ``False``. (:issue:`31503`)
```
Note that it raised only when there were missing values.

TomAugspurger commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -520,7 +520,9 @@ def __getitem__(self, key):
             # first convert to boolean, because check_array_indexer doesn't
             # allow object dtype
-            key = np.asarray(key, dtype=bool)
+            if is_object_dtype(key):
```
Does `is_bool_indexer` imply key is an array, or is it just a list at this point? Ideally we'd do `is_object_dtype(key.dtype)`

dsaxton replied:
`key` could potentially be a list here

jorisvandenbossche replied:
But if it is a list, it is fine to pass it to `check_array_indexer` I think (so something like `[True, False, pd.NA]` works)?

dsaxton replied:
We can pass a list to `check_array_indexer`, but I think the specific case we have to worry about here is when `key` is a numpy array of object dtype, but nonetheless contains bools. In that case it passes `is_bool_indexer` but raises inside `check_array_indexer`, so we explicitly cast it to `bool` dtype here

TomAugspurger commented on pandas/core/indexers.py:
```diff
@@ -333,14 +334,11 @@ def check_array_indexer(array: AnyArrayLike, indexer: Any) -> Any:
 
-    A ValueError is raised when the mask cannot be converted to
-    a bool-dtype ndarray.
+    NA values are treated as False.
```
NA values *in a boolean array*.

In other contexts, they should still raise.

TomAugspurger reviewed (commented):
You'll need to update `boolean.rst`

We should also have a section in `indexing.rst` about this. It'll be important to say that `NA` is treated as False when boolean masking, and raises when present in an integer or string array.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,31 @@ For example:
+
+*pandas 1.1.0*
+
+.. code-block:: python
```
don't use a code-block, rather an ipython block that executes the actual code

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,31 @@ For example:
+
+.. code-block:: python
+
+    >>> s = pd.Series([1, 2, 3, 4])
```
show the setup (e.g. s and mask) in an ipython block above

jreback commented on pandas/core/indexers.py:
```diff
@@ -392,10 +390,10 @@ def check_array_indexer(array: AnyArrayLike, indexer: Any) -> Any:
     dtype = indexer.dtype
     if is_bool_dtype(dtype):
-        try:
+        if is_extension_array_dtype(dtype):
```
we ought to think about adding a ``pd.asarray`` that does exactly this (as we have this idiom in several PRs).

cc @jorisvandenbossche @TomAugspurger

jorisvandenbossche replied:
In general that might be useful, but what would it solve here? (we are converting to numpy, not to an extension array, and we need to use a very specific na_value)

Commit:
Update whatsnew

Commit:
Update boolean.rst

Commit:
check_array_indexer docstring

Commit:
Edit 1.1.0 whatsnew

jorisvandenbossche commented on pandas/core/indexing.py:
```diff
@@ -2207,10 +2208,12 @@ def check_bool_indexer(index: Index, key) -> np.ndarray:
         result = result.astype(bool)._values
-    else:
+    elif is_object_dtype(key):
         # key might be sparse / object-dtype bool, check_array_indexer needs bool array
```
The comment here seems to indicate that this was also done for sparse data (and not only object dtype data). But no tests are failing? This might need some checking if the comment was outdated (I think it was only recently added)

jorisvandenbossche replied:
Can you check this comment? (about the sparse from the code comment)

jorisvandenbossche commented on pandas/tests/arrays/categorical/test_indexing.py:
```diff
@@ -239,15 +239,14 @@ def test_mask_with_boolean(index):
     tm.assert_series_equal(result, expected)
 
 
-@pytest.mark.parametrize("index", [True, False])
```
Can you keep this parametrization?

jorisvandenbossche commented on pandas/tests/extension/base/getitem.py:
```diff
@@ -158,21 +158,21 @@ def test_getitem_boolean_array_mask(self, data):
 
-    def test_getitem_boolean_array_mask_raises(self, data):
+    def test_getitem_boolean_na_treated_as_false(self, data):
         mask = pd.array(np.zeros(data.shape, dtype="bool"), dtype="boolean")
```
Can you take here something with True's as well? (now it will give an empty result)

jorisvandenbossche replied:
it would be good to run the test also for both a boolean array and a list as mask (to ensure the list works)

dsaxton replied:
Looks like the list input may not be working properly, will work on fixing that

dsaxton replied:
@jorisvandenbossche So I think the issue with list masks containing bools and pd.NA was that `is_bool_indexer` from `pandas/core/common.py` didn't consider these to be valid boolean indexers because it was trying to cast to a bool numpy array: https://github.com/pandas-dev/pandas/blob/master/pandas/core/common.py#L142

Made an update there to recognize pd.NA and also updated the test; hopefully CI will still pass. The assumption that boolean indexers are ones that can be cast as _numpy_ boolean arrays seems to happen in a lot of places (e.g., https://github.com/pandas-dev/pandas/blob/master/pandas/core/indexes/base.py#L4147) so I could see this causing problems.

jorisvandenbossche replied:
Hmm, the "old" code of still accepting object dtype makes this a bit more complex indeed. Maybe instead of casting to a numpy array, we could use `pd.array`, so it will handle the case with boolean values better?

dsaxton replied:
I think that should theoretically work in combination with the right change to `is_bool_indexer`

jorisvandenbossche commented on pandas/tests/indexing/test_check_indexer.py:
```diff
@@ -31,15 +31,16 @@ def test_valid_input(indexer, expected):
-    array = np.array([1, 2, 3])
+def test_boolean_na_returns_indexer():
+    # TODO: What to do now about list input that contains
+    # bools and None?
```
Test the behaviour, I would say (and I think it will convert a list to boolean array, so then further treats it as such)

dsaxton replied:
Yes, looks like it's behaving the same way as a boolean array with NA, but I wanted to note this because it might be unexpected to silently treat `None` as `pd.NA` when passed in a list. I'll add it back to the test, which should pass as it's written now.

jorisvandenbossche replied:
Ah, yes, that would indeed be a change in behaviour .. 
I *think* I am fine with that (it's a corner case anyway to pass such mask as a list), but not fully sure. We can also punt for now on the list case (at least for Series, on the array level it should already work)

jorisvandenbossche reviewed (commented):
Thanks for working on this!

Commit:
Add to indexing.rst

Commit:
Add back index parameter

Commit:
Add some True values in test

Commit:
Edit boolean.rst

Commit:
Add list back to check_array_indexer test

Commit:
Merge branch 'master' into bool-idx

Commit:
Account for pd.NA in is_bool_indexer

Commit:
Include list mask in test

Commit:
Account for empty key

jorisvandenbossche commented on pandas/core/common.py:
```diff
@@ -118,31 +118,24 @@ def is_bool_indexer(key: Any) -> bool:
-            return arr.dtype == np.bool_ and len(arr) == len(key)
-        except TypeError:  # pragma: no cover
-            return False
+        return all(k is libmissing.NA or k is True or k is False for k in key) and (
```
Hmm, this might be much less performant for bigger object dtype arrays with only True/False values? 
Also numpy booleans won't pass the identity check I think.

dsaxton replied:
You're right, I'll revert this change

dsaxton replied:
Can you think of an alternative approach that might be more performant? The difficulty seems to be treating _only_ `pd.NA` (not `None` or `NaN`) as a valid missing value, so we seemingly can't just rely on `pd.array(key)` having boolean type, for instance

Commit:
Revert "Account for empty key"

This reverts commit 705947e238c8db258b9df86581368aa68df5260f.

Commit:
Revert "Account for pd.NA in is_bool_indexer"

This reverts commit 578fd3cdf134627013a084e703f6186d019eef75.

jorisvandenbossche commented:
So the discussion I am having with @dsaxton in the inline comments in the following:

Assume we do boolean indexing with a nullable boolean array (or series):

```
s = pd.Series(range(3))
s[pd.array([True, False, pd.NA], dtype="boolean")]   
```

then for consistency, we typically allow to do the same with a list as well:

```
s[[True, False, pd.NA]]
```

(at least that is how indexing now works assuming it would only be True/False values in the list)

However, the question is then: how do we convert such a list to an array? We have `pd.array` for this, but then something like `s[[True, False, None]]` will also be interpreted as nullable booleans, while now this would be an object dtype that will raise an error because you can't index with nulls (and the same for `s[[True, False, np.nan]]`). 

This is because `pd.array([True, False, None])` will infer a boolean array with missing values. 

So:

* This would be a change in behaviour, since indexing with `[True, False, None]` was already possible before the introduction of `pd.NA` and raised an error (although I think we can consider this "mask as list" as a corner case).
* This also interprets None and np.nan as pd.NA, while pd.NA has a different behaviour in those indexing/boolean situations.

Are we OK with that? Or would we like to have a kind of "strict" list conversion, where only actual pd.NA values are seen as missing values?

(whether `pd.array([True, False, np.nan])` should infer boolean and "silently" converts np.nan into pd.NA can maybe also be a discussion)

jorisvandenbossche commented:
(and BTW, I think my comment above discussing this "list" case is rather a corner case, and so I think we can also leave that out for this PR and have that discussion separately, if that makes things easier)

jreback commented:
should not have any special cases in boolean indexing

let’s just go back to the any null value is simply filled False and call it a day
it’s very easy to understand and just works with no downsides

jorisvandenbossche commented:
We already have a special case in boolean indexing, which is that we allow object dtype with boolean values, but raise if those contain missing values. We are not going to change that in this PR. 

The question is how to interpret a list of bools + None/np.nan. Continue interpreting this as object dtype? (and thus raising an error) Or coercing this to nullable  boolean array? (thus changing the indexing behaviour as this PR is introducing)

TomAugspurger commented:
I think just coerce to a nullable boolean array. The rule "list-like inputs are cast with `pd.array(key)` is easy to explain.

Commit:
Try modifying is_bool_indexer

dsaxton commented:
A possible counterargument could be that we'd be making changes to a non-experimental feature (masking with a list of booleans and potentially null values) for the sake of changing an experimental one (masking with a BooleanArray). Also, if we feel that masking with a list is a corner case, that could be seen as an argument for leaving `[True, False, pd.NA]` as-is (even if it's slightly inconsistent with `pd.array([True, False, pd.NA])`), rather than changing the behavior of list masks universally.

I'm mostly concerned because I'm seeing some surprising test failures after making this change (e.g., somehow this of all lines raises: https://github.com/pandas-dev/pandas/blob/master/pandas/tests/indexing/test_loc.py#L968), probably because `is_bool_indexer` is used in a lot of places (or perhaps there's a safer way to make the change).

Commit:
Revert "Try modifying is_bool_indexer"

This reverts commit 8007ce43a5b609cac8ddf46b5f5cc306f6b70eda.

Commit:
Revert "Include list mask in test"

This reverts commit a5593850b97f09fc72e9411851addcd433290b5c.

jorisvandenbossche commented:
As I said before, I am fine with leaving out fixing the list case for now

Commit:
Merge branch 'master' into bool-idx

jreback commented on doc/source/user_guide/boolean.rst:
```diff
@@ -20,8 +20,7 @@ Nullable Boolean data type
 
-pandas does not allow indexing with NA values. Attempting to do so
-will raise a ``ValueError``.
+pandas allows indexing with ``NA`` values in a boolean array, which are treated as ``False``.
```
can you add a versionchanged 1.0.2 here

jreback commented on doc/source/user_guide/indexing.rst:
```diff
@@ -374,6 +374,12 @@ For getting values with a boolean array:
    df1.loc['a'] > 0
    df1.loc[:, df1.loc['a'] > 0]
 
+NA values in a boolean array propogate as ``False``:
```
can you add a versionadded 1.0.2

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,33 @@ For example:
    ser["2014"]
    ser.loc["May 2015"]
 
+Indexing with Nullable Boolean Arrays
```
ok on moving to 1.0.2

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,33 @@ For example:
+    s
+    mask
+
+*pandas 1.0.x*
```
1.0.0-1.0.1

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,33 @@ For example:
+    ...
+    ValueError: cannot mask with array containing NA / NaN values
+
+*pandas 1.1.0*
```
1.0.2

jreback commented on pandas/tests/arrays/categorical/test_indexing.py:
```diff
@@ -240,14 +240,16 @@ def test_mask_with_boolean(index):
 
 @pytest.mark.parametrize("index", [True, False])
-def test_mask_with_boolean_raises(index):
+def test_mask_with_boolean_na_treated_as_false(index):
```
ca you add this issue number here

jreback commented on pandas/tests/extension/base/getitem.py:
```diff
@@ -158,21 +158,22 @@ def test_getitem_boolean_array_mask(self, data):
+    def test_getitem_boolean_na_treated_as_false(self, data):
         mask = pd.array(np.zeros(data.shape, dtype="bool"), dtype="boolean")
         mask[:2] = pd.NA
+        mask[2:4] = True
```
add this issue number (and on any changed tests)

jorisvandenbossche reviewed (commented):
BTW, we also need to update setitem then, in addition to getitem

TomAugspurger commented:
Just to confirm, the expected behavior for setitem is to treat NAs as False?

Commit:
Update release notes and docs

Commit:
Add issue number to tests

dsaxton commented:
> BTW, we also need to update setitem then, in addition to getitem

It looks like `setitem` might just work now after updating `check_bool_indexer` but will add some tests to try to confirm

Commit:
Add some setitem tests

jorisvandenbossche commented:
> Just to confirm, the expected behavior for setitem is to treat NAs as False?

Yes, I would assume so.

> will add some tests to try to confirm

There was just a PR merged that added a bunch of setitem tests, so if you merge latest master, you can update those

Commit:
Revert "Add some setitem tests"

This reverts commit 5a72b2f9e9c99b133a6173e695c77da759fedd74.

Commit:
Merge branch 'master' into bool-idx

Commit:
Update setitem tests

dsaxton commented:
> There was just a PR merged that added a bunch of setitem tests, so if you merge latest master, you can update those

Included a case that replaces False with pd.NA in the setitem tests just added. I think it's safe to simply remove the ones that test only for an error message?

Commit:
Merge branch 'master' into bool-idx

Commit:
Merge branch 'master' into bool-idx

Commit:
Merge branch 'master' into bool-idx

jreback commented on pandas/core/indexing.py:
```diff
@@ -2195,10 +2196,12 @@ def check_bool_indexer(index: Index, key) -> np.ndarray:
         result = np.asarray(result, dtype=bool)
         result = check_array_indexer(index, result)
+    else:
+        result = check_array_indexer(index, result)
```
might be able to simply to check_array_indexer right before returning (iow for all cases) here. (try in a followon)

jreback commented on pandas/tests/extension/test_numpy.py:
```diff
@@ -415,10 +415,6 @@ def test_setitem_mask(self, data, mask, box_in_series):
     def test_setitem_mask_raises(self, data, box_in_series):
         super().test_setitem_mask_raises(data, box_in_series)
 
-    @skip_nested
```
do we have a setitem test?

dsaxton replied:
Yes, I added a case containing NA to the the `test_setitem_mask` test Joris had merged previously (should be line 101 of `setitem.py` above)

jreback commented on pandas/tests/extension/base/setitem.py:
```diff
@@ -124,21 +125,6 @@ def test_setitem_mask_raises(self, data, box_in_series):
         with pytest.raises(IndexError, match="wrong length"):
             data[mask] = data[0]
 
-    def test_setitem_mask_boolean_array_raises(self, data, box_in_series):
```
rather than remove this can you turn it into a test (obviously changed that we no longer raise)

jreback commented:
> > There was just a PR merged that added a bunch of setitem tests, so if you merge latest master, you can update those
> 
> Included a case that replaces False with pd.NA in the setitem tests just added. I think it's safe to simply remove the ones that test only for an error message?

no, simply re-write the tests rather than removing (as indicated)

jreback reviewed (changes requested):
looks good; once the setitem test are good ping on green.

Commit:
Add setitem test

dsaxton commented:
@jreback Seems to be green, the CI failures are due to an apparent web doc issue: https://github.com/pandas-dev/pandas/issues/32060

Commit:
Merge branch 'master' into bool-idx

TomAugspurger commented on pandas/core/indexers.py:
```diff
@@ -419,10 +417,10 @@ def check_array_indexer(array: AnyArrayLike, indexer: Any) -> Any:
+            indexer = indexer.to_numpy(dtype=bool, na_value=False)
+        else:
             indexer = np.asarray(indexer, dtype=bool)
-        except ValueError:
```
What error message do you get now when masking with an object-dtype ndarray containg bools and NaNs?

dsaxton replied:
```python
In [3]: s = pd.Series(range(3))                                                                                                                                                    

In [4]: s                                                                                                                                                                          
Out[4]: 
0    0
1    1
2    2
dtype: int64

In [5]: mask = np.array([True, False, np.nan], dtype=object)                                                                                                                       

In [6]: mask                                                                                                                                                                       
Out[6]: array([True, False, nan], dtype=object)

In [7]: s[mask]                                                                                                                                                                    
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-b9628bf46992> in <module>
----> 1 s[mask]

~/pandas/pandas/core/series.py in __getitem__(self, key)
    877                 key = list(key)
    878 
--> 879             if com.is_bool_indexer(key):
    880                 key = check_bool_indexer(self.index, key)
    881                 return self._get_values(key)

~/pandas/pandas/core/common.py in is_bool_indexer(key)
    128                 na_msg = "Cannot mask with non-boolean array containing NA / NaN values"
    129                 if isna(key).any():
--> 130                     raise ValueError(na_msg)
    131                 return False
    132             return True

ValueError: Cannot mask with non-boolean array containing NA / NaN values
```

TomAugspurger replied:
Good to know, thanks.

jorisvandenbossche commented on pandas/tests/extension/base/setitem.py:
```diff
@@ -124,20 +125,17 @@ def test_setitem_mask_raises(self, data, box_in_series):
 
-    def test_setitem_mask_boolean_array_raises(self, data, box_in_series):
-        # missing values in mask
+    def test_setitem_mask_boolean_array_with_na(self, data, box_in_series):
```
Isn't this test duplicating the `test_setitem_mask` above?

jorisvandenbossche replied:
You can simply remove it then, I think?

jorisvandenbossche commented on doc/source/user_guide/boolean.rst:
```diff
@@ -30,13 +31,6 @@ will raise a ``ValueError``.
    s[mask]
 
-The missing values will need to be explicitly filled with True or False prior
-to using the array as a mask.
```
Can you keep this example but reword it as something like "if you want different behaviour, you can fill manually with fillna(True)" ?

jorisvandenbossche commented on doc/source/whatsnew/v1.0.2.rst:
```diff
@@ -47,6 +47,33 @@ Bug fixes
 .. ---------------------------------------------------------------------------
 
+Indexing with Nullable Boolean Arrays
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Can you move this above bug fixes section, and make this of the same level? (so using `~~~` for underline) 
(it's not a bug fix but a deliberate API change)

jorisvandenbossche reviewed (commented):
Some small comments, thanks for working on this!

dsaxton committed:
Move whatsnew note

dsaxton committed:
Add back example

dsaxton committed:
Merge branch 'master' into bool-idx

dsaxton commented:
@jorisvandenbossche It seems that `check_array_indexer` works okay with sparse input, so maybe the comment is outdated?

```python
In [1]: import numpy as np 
   ...: import pandas as pd 
   ...: from pandas.core.indexing import check_bool_indexer 
   ...: from pandas.core.indexers import check_array_indexer 
   ...:  
   ...: arr = pd.arrays.SparseArray([True, False]) 
   ...: idx = np.array([1, 2]) 
   ...:  
   ...: check_array_indexer(idx, arr) 
   ...:                                                                                                                                                                            
Out[1]: array([ True, False])
```

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into bool-idx

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into bool-idx

jreback reviewed (approved):
lgtm. @jorisvandenbossche

jorisvandenbossche commented:
> It seems that check_array_indexer works okay with sparse input, so maybe the comment is outdated?

That seems so then, yes. Last question: can you update the comment then?

dsaxton committed:
Update comment

jreback commented:
thanks @dsaxton 

very nice!

jreback commented:
@dsaxton 

hmm not sure why this didn’t backport automatically
can u push a PR

dsaxton commented:
> @dsaxton
> 
> hmm not sure why this didn’t backport automatically
> can u push a PR

I'm not too familiar with how backporting works, do I just put a PR against the `1.0.x` branch instead of `master`?

jreback commented:
yes checkout the 1.0.x and cherry pick the commit then push to a new remote branch but before you do that let me see if i can get the bot to do it

jreback commented:
@Meeseeksdev backport to 1.0.x

jreback commented:
@dsaxton ok instructions above

dsaxton commented:
> @dsaxton ok instructions above

Created this PR: https://github.com/pandas-dev/pandas/pull/32192
